src/Shared/FileMatcher.cs

```diff
@@ -1094,215 +1111,280 @@ private static string RegularExpressionFromFileSpec
             out bool isLegalFileSpec
         )
         {
-            isLegalFileSpec = true;
-
-            /*
-             * The code below uses tags in the form <:tag:> to encode special information
-             * while building the regular expression.
-             * 
-             * This format was chosen because it's not a legal form for filespecs. If the
-             * filespec comes in with either "<:" or ":>", return isLegalFileSpec=false to
-             * prevent intrusion into the special processing.
-             */
-            if ((fixedDirectoryPart.IndexOf("<:", StringComparison.Ordinal) != -1) ||
-                (fixedDirectoryPart.IndexOf(":>", StringComparison.Ordinal) != -1) ||
-                (wildcardDirectoryPart.IndexOf("<:", StringComparison.Ordinal) != -1) ||
-                (wildcardDirectoryPart.IndexOf(":>", StringComparison.Ordinal) != -1) ||
-                (filenamePart.IndexOf("<:", StringComparison.Ordinal) != -1) ||
-                (filenamePart.IndexOf(":>", StringComparison.Ordinal) != -1))
+            isLegalFileSpec = IsLegalFileSpec(fixedDirectoryPart, wildcardDirectoryPart, filenamePart);
+            if (!isLegalFileSpec)
             {
-                isLegalFileSpec = false;
-                return String.Empty;
+                return string.Empty;
             }
 
-            /*
-             * Its not legal for there to be a ".." after a wildcard.
-             */
-            if (wildcardDirectoryPart.Contains(dotdot))
+            var matchFileExpression = new StringBuilder(
+                fixedDirectoryPart.Length + wildcardDirectoryPart.Length + filenamePart.Length
+                + FileSpecRegexParts.BeginningOfLine.Length + FileSpecRegexParts.FixedDirWildcardDirSeparator.Length
+                + FileSpecRegexParts.WildcardDirFilenameSeparator.Length + FileSpecRegexParts.EndOfLine.Length
+            );
+            matchFileExpression.Append(FileSpecRegexParts.BeginningOfLine);
+
+            AppendRegularExpressionFromFixedDirectory(matchFileExpression, fixedDirectoryPart);
+            matchFileExpression.Append(FileSpecRegexParts.FixedDirWildcardDirSeparator);
+
+            AppendRegularExpressionFromWildcardDirectory(matchFileExpression, wildcardDirectoryPart);
+            matchFileExpression.Append(FileSpecRegexParts.WildcardDirFilenameSeparator);
```

Comment:
I just wanted to separate the dynamically generated regex parts from the static parts, but the sequence is always the same so it would be fine sticking those in the respective methods.